prov/rstream/src/rstream_ep.c

```diff
@@ -0,0 +1,343 @@
+#include "rstream.h"
+
+static int rstream_ep_close(fid_t fid)
+{
+    int ret;
+    struct rstream_ep *rstream_ep =
+        container_of(fid, struct rstream_ep, util_ep.ep_fid.fid);
+
+    ret = fi_close(&rstream_ep->local_mr.mr->fid);
+    if (ret)
+        return ret;
+    free(rstream_ep->local_mr.base_addr);
+
+    ret = fi_close(&rstream_ep->ep_fd->fid);
+    if (ret)
+        return ret;
+
+    ret = fi_close(&rstream_ep->recv_cq->fid);
+    if (ret)
+        return ret;
+
+    ret = fi_close(&rstream_ep->send_cq->fid);
+    if (ret)
+        return ret;
+
+    ofi_endpoint_close(&rstream_ep->util_ep);
+    free(rstream_ep->tx_ctx.tx_ctxs);
+    free(rstream_ep->rx_ctxs);
+    free(rstream_ep);
+    return 0;
+}
+
+static int rstream_ep_bind(struct fid *ep_fid, struct fid *bfid, uint64_t flags)
+{
+    int ret;
+    struct rstream_ep *rstream_ep =
+        container_of(ep_fid, struct rstream_ep, util_ep.ep_fid.fid);
+    struct rstream_eq *rstream_eq = NULL;
+
+    switch (bfid->fclass) {
+    case FI_CLASS_EQ:
+        rstream_eq = container_of(bfid, struct rstream_eq, util_eq.eq_fid.fid);
+        ret = fi_ep_bind(rstream_ep->ep_fd, &rstream_eq->eq_fd->fid, flags);
+        /* latest bind will always be associated with next cm xchg
+        given rsockets impl. (aka bind then either connect/accept */
+        rstream_eq->bound_ep = rstream_ep;
+	    break;
+    default:
+        FI_WARN(&rstream_prov, FI_LOG_EP_CTRL, "invalid fid class\n");
+        ret = -FI_EINVAL;
+        break;
+    }
+    return ret;
+}
+
+static int rstream_ep_ctrl(struct fid *fid, int command, void *arg)
+{
+    struct rstream_ep *rstream_ep;
+    int ret;
+    rstream_ep = container_of(fid, struct rstream_ep, util_ep.ep_fid.fid);
+
+    switch (command) {
+    case FI_ENABLE:
+        ret = fi_enable(rstream_ep->ep_fd);
+        break;
+    case FI_GETWAIT: /* need to have one cq->fd */
+        ret = fi_control(&rstream_ep->send_cq->fid, FI_GETWAIT, arg);
+        break;
+    default:
+        return -FI_ENOSYS;
+    }
+    return ret;
+}
+
+static struct fi_ops rstream_ep_fi_ops = {
+    .size = sizeof(struct fi_ops),
+    .close = rstream_ep_close,
+    .bind = rstream_ep_bind,
+    .control = rstream_ep_ctrl,
+    .ops_open = fi_no_ops_open,
+};
+
+static struct fi_ops_ep rstream_ops_ep = {
+    .size = sizeof(struct fi_ops_ep),
+    .cancel = fi_no_cancel,
+    .getopt = fi_no_getopt,
+    .setopt = fi_no_setopt,
+    .tx_ctx = fi_no_tx_ctx,
+    .rx_ctx = fi_no_rx_ctx,
+    .rx_size_left = fi_no_rx_size_left,
+    .tx_size_left = fi_no_tx_size_left,
+};
+
+static int rstream_reg_mrs(struct fid_domain *domain,
+    struct rstream_lmr_data *lmr)
+{
+    int ret;
+    uint64_t rx_meta_data_offset = 0;
+    uint32_t full_mr_size = RSTREAM_MAX_MR_SEG_SIZE * 2;
+
+    if (RSTREAM_USING_IWARP)
+        rx_meta_data_offset = RSTREAM_IWARP_DATA_SIZE * RSTREAM_MAX_RX_OPS;
```

Comment:
The transmit and receive buffer sizes must be configurable, and able to be modified by the app.

Reply:
what would you be your recommendation on how to tell the provider this info through OFI when a user sets this sockopt? max_msg_size in ep_attr isn't able to differentiate between tx/rx. I don't see a related setting for tx/rx_attr or on the domain_attr.

Reply:
max_msg_size doesn't apply very well to a data stream.  I would use a new control operation to get/set this value, which maps to how socket handles it.